File: PathSum2BruteForce.java
// Time complexity: O(n^2) n for traversing the nodes and n for deep copying to a new list everytime
// Space complexity: O(n*H) recursive stack times a new list for every call

// Approach: Brute force. We add to the curr sum at every node and store the sum, and pathnodes as a new list at every stack element. We don't use the same list as it is just a pointer to the list. So at the end our list will contain all elements instead of the actual path that summed up to the target. To overcome this, we use a new list for every function call

class Solution {
    List<List<Integer>> result = new ArrayList<>();
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        helper(root, targetSum, 0, new ArrayList<Integer>());
        return result;
    }
    
    private void helper(TreeNode root, int targetSum, int currSum, List<Integer> pathNodes) {
        // base
        if (root == null) return;
        
        // action
        pathNodes.add(root.val);
        
        // logic
        currSum += root.val;
        if (root.left == null && root.right == null) {
            if (currSum == targetSum) {
                result.add(pathNodes);
            }
        }
        helper(root.left, targetSum, currSum, new ArrayList<>(pathNodes));
        helper(root.right, targetSum, currSum, new ArrayList<>(pathNodes));
    }
}

File: path_sum2_bruteforce.py
"""
Time Complexity : O(n^2) where n is the number of nodes in the tree. O(n) required for visiting every tree node and additional o(n) is required for deep copy
Space Complexity : O(n^2) where n is the number of nodes in the tree. O(n) for recursive stack and additional O(n) for the list created for each node

Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No

// Your code here along with comments explaining your approach
1. Data structure inside a data structure is a reference.
2. So if we do not create a separate list for every node, at the end the all the elements will be preseneted in the result list as we are making changes to the same reference/list
3. For every node we will maintain three local things, root, currentSum and its associated path
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def __init__(self):
        self.result = []

    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        if root == None:
            return self.result
        self.helper(root, 0, [], targetSum)
        return self.result

    def helper(self, root, current_sum, path, target):
        # base
        if root == None:
            return
        if root.left == None and root.right == None:
            if current_sum + root.val == target:
                path.append(root.val)
                self.result.append(path)
                return
        # logic
        current_sum += root.val
        path.append(root.val)
        self.helper(root.left, current_sum, path[:], target)
        self.helper(root.right, current_sum, path[:], target)
